### ISSUE TITLE
Figure out how to handle immutability in Lenses

### ISSUE THREAD

**Issue by 5outh** [5outh/nanoscope#63]
Right now, nanoscope uses `_.cloneDeep` from lodash to deep clone objects for immutability's sake. There are a number of downsides to this:

- It does a full clone on every set/map call, and is slow
- Sometimes users may want or expect mutability when calling something like `set`
- This is concerning, from the lodash docs:

> An empty object is returned for uncloneable values such as functions, DOM nodes, Maps, Sets, and WeakMaps.

I'm thinking it would be nice to implement a cloning strategy as an option for `nanoscope`, that is one of `mutable` or `immutable`, where the latter uses immutable-js instead of lodash.

i.e `cloneDeep` should disappear and be replaced by immutable-js, and the user should be able to specify that they want to directly mutate by setting an option.

Since pulling in immutable-js as a dependency is somewhat heavy-handed, this feature is definitely up for discussion!

**Comment by hallettj** [comment]
Is it possible to implement lenses that operate on Immutable.js values without requiring Immutable as a dependency?

My preference is that setters produce the same type of data that they take in. E.g., updating a vanilla object produces a new vanilla object, updating an Immutable Record produces a new Record. That lets users who want Immutable values to opt in by using them from the start.

I mentioned in my pull request that I would prefer shallow clones with updating mutable structures - even if that means weakening data isolation guarantees. Part of my motivation is that I would like to be able to work with structures such as a vanilla object that contains Immutable Lists. I'm trying to get type safety with Flow, and type-checking Immutable's Records is tricky.

I don't want to update values in place. My ideal lens creates new objects when updating a property on that object (or its descendent) where unmodified properties of the new object are references to the same objects that the original object referenced. (I think this is how React's immutability helpers operate, for example.)

In the past when I wanted to update an object without modifying the original, I would use lodash in this manner:

    var updated = _.assign({}, original, changes)

In conclusion, my preference is to add or extend lenses to operate on Immutable.js values if they are given as input, to avoid adding Immutable.js as a hard dependency, and to modify existing behavior to make shallow clones.

**Comment by 5outh** [comment]
Great, thanks for your input! :+1:
